Sources/PackageGraph/CheckoutState.swift

```diff
@@ -15,50 +15,44 @@ import TSCUtility
 /// A checkout state represents the current state of a repository.
 ///
 /// A state will always has a revision. It can also have a branch or a version but not both.
-public struct CheckoutState: Equatable, Hashable {
+public enum CheckoutState: Equatable, Hashable {
 
-    /// The revision of the checkout.
-    public let revision: Revision
-
-    /// The version of the checkout, if known.
-    public let version: Version?
-
-    /// The branch of the checkout, if known.
-    public let branch: String?
-
-    /// Create a checkout state with given data. It is invalid to provide both version and branch.
-    ///
-    /// This is deliberately fileprivate so CheckoutState is not initialized
-    /// with both version and branch. All other initializers should delegate to it.
-    fileprivate init(revision: Revision, version: Version?, branch: String?) {
-        assert(version == nil || branch == nil, "Can't set both branch and version.")
-        self.revision = revision
-        self.version = version
-        self.branch = branch
-    }
-
-    /// Create a checkout state with given revision and branch.
-    public init(revision: Revision, branch: String? = nil) {
-        self.init(revision: revision, version: nil, branch: branch)
-    }
+    case revision(Revision)
+    case version(Version, revision: Revision)
+    case branch(String, revision: Revision)
```

Comment:
Should we name the first parameter, too?

Reply:
I started with naming them but it was a bit awkward in the call site e.g. `let checkoutState: CheckoutState = .branch(branch: "", revision: "")`

Reply:
Right, the first parameter is often unnamed when the enum case makes it obvious, and I agree that that applies here.  So I think it should be unnamed.

Reply:
ended up adding `name: ` to the branch variant